schema/validator.go

```diff
@@ -117,3 +119,66 @@ func validateManifestDescendants(r io.Reader) error {
 	}
 	return nil
 }
+
+func validateManifestList(r io.Reader) error {
+	header := v1.ManifestList{}
+
+	buf, err := ioutil.ReadAll(r)
+	if err != nil {
+		return errors.Wrapf(err, "error reading the io stream")
+	}
+
+	err = json.Unmarshal(buf, &header)
+	if err != nil {
+		return errors.Wrap(err, "manifestlist format mismatch")
+	}
+
+	for _, manifest := range header.Manifests {
+		checkPlatform(manifest.Platform.OS, manifest.Platform.Architecture)
+	}
+
+	return nil
+}
+
+func validateConfig(r io.Reader) error {
+	header := v1.Image{}
+
+	buf, err := ioutil.ReadAll(r)
+	if err != nil {
+		return errors.Wrapf(err, "error reading the io stream")
+	}
+
+	err = json.Unmarshal(buf, &header)
+	if err != nil {
+		return errors.Wrap(err, "config format mismatch")
+	}
+
+	checkPlatform(header.OS, header.Architecture)
+
+	return nil
+}
+
+func checkPlatform(OS string, Architecture string) {
+	validCombins := map[string][]string{
+		"android":   {"arm"},
+		"darwin":    {"386", "amd64", "arm", "arm64"},
+		"dragonfly": {"amd64"},
+		"freebsd":   {"386", "amd64", "arm"},
+		"linux":     {"386", "amd64", "arm", "arm64", "ppc64", "ppc64le", "mips64", "mips64le", "s390x"},
+		"netbsd":    {"386", "amd64", "arm"},
+		"openbsd":   {"386", "amd64", "arm"},
+		"plan9":     {"386", "amd64"},
+		"solaris":   {"amd64"},
+		"windows":   {"386", "amd64"}}
+	for os, archs := range validCombins {
+		if os == OS {
+			for _, arch := range archs {
+				if arch == Architecture {
+					break
+				}
+			}
+			fmt.Printf("warning: combination of %q and %q is invalid.", OS, Architecture)
+		}
+	}
+	fmt.Printf("warning: operating system %q of the bundle is not supported yet.", OS)
```

Comment:
> I am almost ready to propose a moratorium on validation until someone starts doing some engineering work.

We just need a way to handle warnings for SHOULD violations (they aren't errors, but the caller may want to know about them).  I've tried floating `strict` (#403, opencontainers/image-tools#66) and TAP comments (opencontainers/image-tools#60).  And @xiekeyang has been trying to follow your guidance on this in #452.  So far nobody has managed to land anything more polished than “print warnings to stdout” (and we have [precedence for that approach in master][1]).  So until the image-spec maintainers decide how they want SHOULD violations handled, I think we should just roll forward with `fmt.Printf("warning:…`.  It will be easy enough to grep and replace once we have a better approach figured out.

[1]: https://github.com/opencontainers/image-spec/blob/ed18de10b0d7e34076e1557bdcd6f500504a6673/schema/validator.go#L109